File: platform/windows/Instance.cs
﻿using System;
using System.Text;

namespace FullStacked
{
    internal class Instance
    {
        public String id;
        public Boolean isEditor;

        private byte[] header;

        public Instance(String id, Boolean isEditor = false) {
            this.id = id;
            this.isEditor = isEditor;

            if (isEditor)
            {
                this.header = new byte[] { 1 }; // isEditor
                this.header = App.combineBuffers([this.header, App.numberToByte(0)]); // no project id
            }
            else { 
                this.header = new byte[] { 0 };
                byte[] idData = Encoding.UTF8.GetBytes(id);
                this.header = App.combineBuffers([this.header, App.numberToByte(idData.Length)]);
                this.header = App.combineBuffers([this.header, idData]);
            }
        }

        public byte[] callLib(byte[] payload) {
            byte[] data = App.combineBuffers([this.header, payload]);

            return App.call(data);
        }

    }
}

File: platform/apple/FullStacked/Instance.swift
import SwiftUI

class Instance {
    static var callId = 0
    
    public let isEditor: Bool
    public let id: String
    private var header: Data
    
    
    init (projectId: String, isEditor: Bool = false) {
        self.isEditor = isEditor
        self.id = projectId
        
        self.header = Data()
        if(isEditor) {
            self.header.append(Data([1])) // isEditor
            self.header.append(0.toBytes()) // no project id
        } else {
            self.header.append(Data([0]))
            let projectIdData = self.id.data(using: .utf8)!
            self.header.append(projectIdData.count.toBytes())
            self.header.append(projectIdData)
        }
    }
    
    func callLib(payload: Data) -> Data {
        var data = Data()
        data.append(self.header)
        data.append(payload)
        
        let id = Int32(Instance.callId);
        let size = call(id, data.ptr(), Int32(data.count))
        
        let responsePtr = UnsafeMutableRawPointer.allocate(byteCount: Int(size), alignment: 1)
        
        getResponse(id, responsePtr)
        
        let response = Data(bytes: responsePtr, count: Int(size))
        responsePtr.deallocate()
        
        Instance.callId += 1;
        
        return response
    }
}
